packages/arb-validator-core/gobridge/goEth.go

```diff
@@ -0,0 +1,370 @@
+	m.nextAddress = common.BigIntToAddress(addrInt.Add(addrInt, big.NewInt(1)))
+	return addr
+}
+
+func (m *goEthdata) getCurrentBlock() *common.BlockId {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	return m.NextBlock
+}
+
+func (m *goEthdata) getLastBlock() *common.BlockId {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	return m.LastMinedBlock
+}
+
+func (m *goEthdata) getBlockFromHeight(height *common.TimeBlocks) (*common.BlockId, error) {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	b, ok := m.blockNumbers[height.AsInt().Uint64()]
+	if !ok {
+		return nil, errors.New("not found")
+	}
+
+	return b, nil
+}
+
+func (m *goEthdata) registerOutChan(oc chan arbbridge.MaybeEvent) {
+	m.chanMgr <- oc
+}
+
+func (m *goEthdata) pubMsg(challenge *challengeData, msg arbbridge.MaybeEvent) {
+	m.msgMutex.Lock()
+	defer m.msgMutex.Unlock()
+	m.nextMsgs[msg.Event.GetChainInfo().BlockId] = append(m.nextMsgs[msg.Event.GetChainInfo().BlockId], goMaybeEvent{
+		challenge: challenge,
+		event:     msg,
+	})
+}
+
+func mine(m *goEthdata, t time.Time) {
+	m.blockMutex.Lock()
+	blockHash := hashing.SoliditySHA3(hashing.Uint256(big.NewInt(rand2.Int63())))
+	m.NextBlock.HeaderHash = blockHash
+	lastBlock := m.LastMinedBlock
+	m.blockNumbers[m.NextBlock.Height.AsInt().Uint64()] = m.NextBlock
+	m.blockHashes[m.NextBlock.HeaderHash] = m.NextBlock
+	m.parentHashes[*m.NextBlock] = lastBlock.HeaderHash
+	m.LastMinedBlock = m.NextBlock
+	m.msgMutex.Lock()
+	for _, event := range m.nextMsgs[m.NextBlock] {
+		for _, ru := range m.rollups {
+			ru.events[m.NextBlock] = append(ru.events[m.NextBlock], event.event.Event)
+		}
+		m.challengeWatchersMutex.Lock()
+		for challenge, watcher := range m.challengeWatcherEvents {
+			// if this is a challenge specific event and it is not for this challenge ignore it
+			if event.challenge != nil && event.challenge != challenge {
+				continue
+			}
+			watcher[m.NextBlock] = append(watcher[m.NextBlock], event.event.Event)
+		}
+		m.challengeWatchersMutex.Unlock()
+		m.pubchan <- event.event
+	}
+	m.msgMutex.Unlock()
+	log.Println("mined block number", m.NextBlock)
+	newBlock := new(common.BlockId)
+	newBlock.Height = common.NewTimeBlocks(new(big.Int).Add(m.LastMinedBlock.Height.AsInt(), big.NewInt(1)))
+	m.NextBlock = newBlock
+	blockEvent := arbbridge.NewTimeEvent{
+		arbbridge.ChainInfo{
+			BlockId: m.LastMinedBlock,
+		},
+	}
+
+	m.blockMutex.Unlock()
+	m.pubMsg(nil, arbbridge.MaybeEvent{
+		Event: blockEvent,
+	})
+}
+
+func addToken(source common.Address, data value.Value, destination common.Address, amount *big.Int) {
+	if amount.Cmp(big.NewInt(0)) == 0 {
+		return
+	}
+
+}
+
+func (m *goEthdata) deliverMessage(address common.Address, msgHash common.Hash) {
+	hash := hashing.SoliditySHA3(
+		hashing.Bytes32(m.inbox[address].value),
+		hashing.Bytes32(msgHash),
+	)
+	m.inbox[address].value = hash
+	m.inbox[address].count = new(big.Int).Add(m.inbox[address].count, big.NewInt(1))
+}
```

Comment:
This should be part of the `GlobalInbox`

Reply:
fixed